src/main/java/uk/gov/companieshouse/web/accounts/util/ValidationHelper.java

```diff
@@ -0,0 +1,200 @@
+import uk.gov.companieshouse.api.error.ApiError;
+import uk.gov.companieshouse.api.error.ApiErrorResponseException;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+import uk.gov.companieshouse.web.accounts.enumeration.ValidationMessage;
+import uk.gov.companieshouse.web.accounts.exception.MissingValidationMappingException;
+import uk.gov.companieshouse.web.accounts.validation.ValidationMapping;
+import uk.gov.companieshouse.web.accounts.validation.ValidationModel;
+import uk.gov.companieshouse.web.accounts.validation.ValidationError;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static uk.gov.companieshouse.web.accounts.CompanyAccountsWebApplication.APPLICATION_NAME_SPACE;
+
+/**
+ * The {@code ValidationHelper} class provides support methods for handling
+ * API validation errors and makes use of the {@link ValidationModel} and
+ * {@link ValidationMapping} annotations.
+ * <p>
+ * Top-level models should be annotated with {@code ValidationModel}.
+ * Model fields associated with specific API validation errors should be
+ * annotated with {@code ValidationMapping} and include the full JSON error
+ * path that will be returned by the API.
+ * <p>
+ * The {@code scanPackageForValidationMappings} method must be called at
+ * startup to initialise the list of mappings between the API JSON error
+ * paths and model fields that other methods of this class depend on.
+ *
+ * @see ValidationModel
+ * @see ValidationMapping
+ * @see ValidationMessage
+ */
+@Component
+public class ValidationHelper {
+
+    private static final int MAX_DEPTH = 5;
+
+    private static int depth = 0;
+
+    private static HashMap<String, String> mappings;
+
+    private static final String PATH_DELIMITER = ".";
+
+    private static final Set<Class<?>> primitivesSet = createPrimitivesSet();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(APPLICATION_NAME_SPACE);
+
+    private ValidationHelper() {}
+
+    /**
+     * Scans classes belonging to the specified package for any candidate
+     * classes annotated with the {@code ValidationModel} annotation, and
+     * generates a collection of mappings of JSON error paths to model object
+     * fields to support the handling of API validation errors.
+     *
+     * @param scanner {@link ClassPathScanningCandidateComponentProvider}
+     * @param basePackage       the base package from which to begin scanning
+     * @see   ValidationModel
+     * @see   ValidationMapping
+     */
+    public static void scanPackageForValidationMappings(ClassPathScanningCandidateComponentProvider scanner, String basePackage) {
+
+        scanner.addIncludeFilter(new AnnotationTypeFilter(ValidationModel.class));
+
+        mappings = new HashMap<>();
+        depth = 0;
```

Comment:
is this thread safe?

Reply:
As @BSaunders95 has suggested, the scan is executed once on application startup from the main thread, so this shouldn't be a concern. In general, there are no guarantees as to the thread safety of methods in this class—if we change the way validation mappings are scanned in future this may need to be considered.